docs/dev/testing.rst

```diff
@@ -0,0 +1,306 @@
+
+.. code-block:: console
+
+   $ pytest test/unit/extract/excel_test.py
+
+To run only the unit tests defined by a single test class within that module:
+
+.. code-block:: console
+
+   $ pytest test/unit/extract/excel_test.py::TestGenericExtractor
+
+Custom PUDL pytest flags
+^^^^^^^^^^^^^^^^^^^^^^^^
+We have defined several custom flags to control pytest's behavior when running
+the PUDL tests. You can always check to see what custom flags exist by running
+``pytest --help`` and looking at the ``custom options`` section:
+
+.. code-block:: console
+
+  custom options:
+  --live-dbs            Use existing PUDL/FERC1 DBs instead of creating temporary ones.
+  --tmp-data            Download fresh input data for use with this test run only.
+  --etl-settings=ETL_SETTINGS
+                        Path to a non-standard ETL settings file to use.
+  --gcs-cache-path=GCS_CACHE_PATH
+                        If set, use this GCS path as a datastore cache layer.
+  --sandbox             Use raw inputs from the Zenodo sandbox server.
+
+Run the output tests using a pre-existing PUDL database rather than building
+a new one for testing:
+
+.. code-block:: console
+
+  $ pytest --live-dbs test/integration/fast_output_test.py
+
+Run the ETL portion of the integration tests, and force it to download fresh
+input data to a temporary datastore:
+
+.. code-block:: console
+
+   $ pytest --tmp-data test/integration/etl_test.py
+
+Run the ETL portion of the integration tests, but using a non-standard settings
+file (equivalent to ``tox -e ferc1_solo``):
+
+.. code-block:: console
+
+  $ pytest --etl-settings=test/settings/ferc1-solo-test.yml test/integration/etl_test.py
+
+Check that the schemas of all historical FERC 1 databases are compatible with
+the most recent year, which we use to generate the schema for our FERC 1 SQLite
+database containing all years of FERC 1 data (equivalent to ``tox -e ferc1_schema``):
+
+.. code-block:: console
+
+  $ pytest --etl-settings test/settings/full-integration-test.yml test/integration/etl_test.py::test_ferc1_schema
+
+-------------------------------------------------------------------------------
+Data Validation
+-------------------------------------------------------------------------------
+Once the PUDL ETL pipeline has processed all of the available data, we have a
+collection of tests that can be run to verify a collection of expectations
+about what the outputs should look like. These data validation tests are
+organized into datasource specific modules under ``test/validate``. Running
+the full data validation can take as much as an hour, depending on your
+computer.
+
+These tests require a fully populated PUDL database which contains all
+available FERC and EIA data, as specified by the
+``test/settings/full-integration-test.yml`` input file. They are run against the
+"live" SQLite database in your pudl workspace at ``sqlite/pudl.sqlite``. To run
```

Comment:
i may have missed this but do we still have an explanation of the pudl_setup somewhere?

Reply:
Hmm. Where should that live at this point? Should it be user-facing at all? Or should it only be part of the developer docs? If the latter, I guess it should be in `dev_setup.rst` -- if it's not already. Which it might not be, since it used to be in the usage / user setup section.

Reply:
definitely should be part of developer docs!

Reply:
I see that I am referring to the Usage section for this inside the developer docs.  I guess that's not actually where it needs to be, since anyone who is accessing the data within a Docker container won't have to worry about the workspace setup.

Reply:
yeah exactly. we had talked about nesting some things to conform to [this outline](https://docs.google.com/document/d/1hBWSpBARpr27rcR4ongK4RgZ-Fhi0T99IpuMwvtUPyY/edit) that we agreed upon.